packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart

```diff
@@ -109,26 +77,168 @@ class FuchsiaPM {
-      printError('Error archiving Fuchsia package: ');
-      printError(result.stdout);
-      printError(result.stderr);
+    final Process process = await runCommand(command);
+    process.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(printTrace);
+    process.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(printError);
+    return process;
+  }
+
+  /// Publishes a Fuchsia package to a served package repository.
+  ///
+  /// For a package repo initialized with [newrepo] at [repoPath] and served
+  /// by [serve], this call publishes the `far` package at [packagePath] to
+  /// the repo such that it will be visible to devices connecting to the
+  /// package server.
+  Future<bool> publish(String repoPath, String packagePath) {
+    return _runPMCommand(<String>[
+      'publish',
+      '-a',
+      '-r',
+      repoPath,
+      '-f',
+      packagePath,
+    ]);
+  }
+
+  Future<bool> _runPMCommand(List<String> args) async {
+    if (fuchsiaArtifacts.pm == null) {
+      throwToolExit('Fuchsia pm tool not found');
+    }
+    final List<String> command = <String>[fuchsiaArtifacts.pm.path] + args;
+    final RunResult result = await runAsync(command);
+    return result.exitCode == 0;
+  }
+}
+
+/// A class for running and retaining state for a Fuchsia package server.
+///
+/// [FuchsiaPackageServer] takes care of initializing the package repository,
+/// spinning up the package server, publishing packages, and shutting down the
+/// the server.
+///
+/// Example usage:
+/// var server = FuchsiaPackageServer(
+///     '/path/to/repo',
+///     await FuchsiaDevFinder.resolve(deviceName),
+///     await freshPort());
+/// try {
+///   await server.start();
+///   await server.addPackage(farArchivePath);
+///   ...
+/// } finally {
+///   server.stop();
+/// }
+class FuchsiaPackageServer {
+  FuchsiaPackageServer(
+    String repo,
+    String host,
+    int port
+  ) : _repo = repo, _host = host, _port = port;
+
+  final String _repo;
+  final String _host;
+  final int _port;
+
+  Process _process;
+
+  /// The url that can be used by the device to access this package server.
+  String get url => 'http://$_host:$_port';
+
+  /// Usees [FuchiaPM.newrepo] and [FuchsiaPM.serve] to spin up a new Fuchsia
+  /// package server.
+  ///
+  /// Returns false if ther repo could not be created or the server could not
+  /// be spawned, and true otherwise.
+  Future<bool> start() async {
+    if (_process != null) {
+      printError('$this already started!');
+      return false;
+    }
+    // initialize a new repo.
+    if (!await fuchsiaSdk.fuchsiaPM.newrepo(_repo)) {
+      printError('Failed to create a new package server repo');
       return false;
     }
+    _process = await fuchsiaSdk.fuchsiaPM.serve(_repo, _host, _port);
+    // Put a completer on _process.exitCode to watch for error.
+    unawaited(_process.exitCode.then((int c) {
```

Comment:
I think `whenComplete` doesn't require the unawaited call

Reply:
It also requires unawaited =(